File: extensions/effekseer/assets/framework/efk_model.ts
import { director, gfx, Material, renderer, RenderingSubMesh } from "cc";

const ModelVertexEx = [                                         //                  = 72
    new gfx.Attribute("Input_Pos", gfx.Format.RGB32F),          // Pos      -> vec3 + 12   
    new gfx.Attribute("Input_Normal", gfx.Format.RGB32F),       // Normal   -> vec3 + 12
    new gfx.Attribute("Input_Binormal", gfx.Format.RGB32F),     // Binormal -> vec3 + 12
    new gfx.Attribute("Input_Tangent", gfx.Format.RGB32F),      // Tangent  -> vec3 + 12
    new gfx.Attribute("Input_UV", gfx.Format.RG32F),            // UV1      -> vec2 + 8
    new gfx.Attribute("Input_Color", gfx.Format.RGBA32F),       // Color    -> vec4 + 16
];

const ModelStrip = 72;

// 特效 自定义模型渲染

const g_models: Record<number, {
    vertex: gfx.Buffer,
    indice: gfx.Buffer,
    ia: gfx.IndirectBuffer
}[]> = {};

export class EFKModel {
    private static m_meshs: RenderingSubMesh[] = [];
    private static m_typeid_meshs: Record<number, number[]> = {};
    private static m_typeid_used: Record<number, boolean> = {};

    private static m_model_meshs: Record<number, number[]> = {};
    private static m_model_cache: Record<number, number[]> = {};

    private static m_models: renderer.scene.Model[] = [];
    private static m_model_used: Record<number, boolean> = {};

    // priority 决定渲染顺序
    private static m_priority: number = 0;

    public static doDestroy() {
        EFKModel.m_meshs.forEach(mesh => {
            mesh.vertexBuffers.length = 0;
            mesh.destroy();
        })
        EFKModel.m_meshs = [];
    }


    // 特效 sprite 网格数据渲染

    private static m_buffers: Record<number, {
        vertex: gfx.Buffer,
        indice: gfx.Buffer
    }> = {};

    public static updateMesh(stride: number, vertecies: Uint8Array) {
        const device = gfx.deviceManager.gfxDevice;

        const buffer = EFKModel.m_buffers[stride];
        if (!buffer || buffer.vertex.size < vertecies.byteLength) {
            const vertexBuffer = buffer ? buffer.vertex : device.createBuffer(new gfx.BufferInfo(
                gfx.BufferUsageBit.VERTEX | gfx.BufferUsageBit.TRANSFER_DST,
                gfx.MemoryUsageBit.HOST | gfx.MemoryUsageBit.DEVICE,
                vertecies.byteLength,
                stride,
            ));

            if (buffer) {
                vertexBuffer.resize(vertecies.byteLength);
            }

            // 顶点数/4 = 多少四边形，四边形由 2 个三角形组成，三角形由 3 个顶点组成
            const indexCount = 3; // (vertexBuffer.size / stride) / 4 * 2 * 3;
            const indexBuffer: gfx.Buffer = buffer ? buffer.indice : device.createBuffer(new gfx.BufferInfo(
                gfx.BufferUsageBit.INDEX | gfx.BufferUsageBit.TRANSFER_DST,
                gfx.MemoryUsageBit.DEVICE,
                indexCount * Uint16Array.BYTES_PER_ELEMENT,
                Uint16Array.BYTES_PER_ELEMENT,
            ));

            EFKModel.m_buffers[stride] = {
                vertex: vertexBuffer,
                indice: indexBuffer
            }
        }

        EFKModel.m_buffers[stride].vertex.update(vertecies);
    }

    protected static resizeIndice(indexBuffer: gfx.Buffer, indexCount: number) {
        if (indexBuffer.size >= indexCount * Uint16Array.BYTES_PER_ELEMENT) {
            return;
        }

        indexBuffer.resize(indexCount * Uint16Array.BYTES_PER_ELEMENT);
        const indices: Uint16Array = new Uint16Array(indexCount);

        let element = 0;
        for (let i = 0; i < indexCount;) {
            const baseIdx = 4 * element;
            indices[i++] = baseIdx;
            indices[i++] = baseIdx + 1;
            indices[i++] = baseIdx + 2;
            indices[i++] = baseIdx + 3;
            indices[i++] = baseIdx + 2;
            indices[i++] = baseIdx + 1;
            element++;
        }
        indexBuffer.update(indices);
    }

    public static drawSprite(stride: number, offset: number, spriteCount: number, material: renderer.MaterialInstance, typeid: number): renderer.scene.Model {
        const fisrtVertex = offset / stride;
        const fisrtIndex = fisrtVertex / 4 * 6;

        const buffer = EFKModel.m_buffers[stride];
        if (null == buffer) debugger

        // 索引偏移 + 索引数
        EFKModel.resizeIndice(buffer.indice, fisrtIndex + spriteCount * 6);

        const iaInfo = new gfx.DrawInfo(
            spriteCount * 4,
            fisrtVertex,
            spriteCount * 6,
            fisrtIndex
        );

        let mesh = -1;
        let meshs = EFKModel.m_typeid_meshs[typeid] = EFKModel.m_typeid_meshs[typeid] || [];
        for (let i = 0; i < meshs.length; ++i) {
            let idx = meshs[i];
            if (!EFKModel.m_typeid_used[idx]) {
                EFKModel.m_typeid_used[idx] = true;
                mesh = idx;
                break;
            }
        }

        let subMesh: RenderingSubMesh = null!;
        if (mesh >= 0) {
            subMesh = EFKModel.m_meshs[mesh];
            subMesh.drawInfo = iaInfo;
        }
        else {
            const attrs: gfx.Attribute[] = [];
            material.passes[0].shaderInfo.attributes.forEach(attr => {
                attrs.push(new gfx.Attribute(attr.name, attr.format, attr.isNormalized, attr.stream, attr.isInstanced, attr.location));
            })
            subMesh = new RenderingSubMesh([buffer.vertex], attrs, gfx.PrimitiveMode.TRIANGLE_LIST, buffer.indice);

            mesh = EFKModel.m_meshs.length;
            EFKModel.m_meshs.push(subMesh);

            EFKModel.m_typeid_meshs[typeid].push(mesh);
            EFKModel.m_typeid_used[mesh] = true;
        }

        return EFKModel.checkModel(mesh, subMesh, material, iaInfo);
    }

    static sum_count: number = 0;
    public static addModel(key: number, vertecies: Uint8Array, indices: Uint8Array, frame: number) {
         const device = gfx.deviceManager.gfxDevice;
        g_models[key] = g_models[key] || [];

        const buffer = g_models[key][frame];
        if (!buffer || buffer.vertex.size < vertecies.byteLength) {
            const vertexBuffer = buffer ? buffer.vertex : device.createBuffer(new gfx.BufferInfo(
                gfx.BufferUsageBit.VERTEX | gfx.BufferUsageBit.TRANSFER_DST,
                gfx.MemoryUsageBit.HOST | gfx.MemoryUsageBit.DEVICE,
                vertecies.byteLength,
                ModelStrip
            ));

            const indexBuffer: gfx.Buffer = buffer ? buffer.indice : device.createBuffer(new gfx.BufferInfo(
                gfx.BufferUsageBit.INDEX | gfx.BufferUsageBit.TRANSFER_DST,
                gfx.MemoryUsageBit.DEVICE,
                indices.byteLength,
                Int32Array.BYTES_PER_ELEMENT,
            ));

            if (buffer) {
                vertexBuffer.resize(vertecies.byteLength);
                indexBuffer.resize(indices.byteLength);
            }

            g_models[key][frame] = {
                vertex: vertexBuffer,
                indice: indexBuffer,
                ia: new gfx.IndirectBuffer([new gfx.DrawInfo(
                    vertexBuffer.count, 0, indexBuffer.count, 0
                )])
            }

            //this.sum_count += (vertecies.byteLength + indices.byteLength)/1024/1024;
            //console.log(key, frame, (this.sum_count).toFixed(2), " mb");
        }

        g_models[key][frame].vertex.update(vertecies);
        g_models[key][frame].indice.update(indices);
    }

    public static removeModel(key: number) {        
        if (null == g_models[key]) {
            return;
        }

        // 清除 buffer 对象
        g_models[key].forEach(frame => {
            frame.vertex.destroy();
            frame.indice.destroy();
        });

        delete g_models[key];

        EFKModel.m_model_meshs[key].forEach(mesh => {
            // 清除 mesh 对象
            EFKModel.m_meshs[mesh].destroy();
            delete EFKModel.m_meshs[mesh];

            // 清除 model 对象            
            let models = EFKModel.m_model_cache[mesh];
            models.forEach(m => {
                EFKModel.m_models[m].destroy();
                delete EFKModel.m_models[m];
            });
        });
        delete EFKModel.m_model_meshs[key];
    }

    public static drawModel(id: number, frame: number, material: renderer.MaterialInstance, typeid: number): renderer.scene.Model {
        // return;
        let meshs = EFKModel.m_model_meshs[id] = EFKModel.m_model_meshs[id] || [];
        let mesh = (meshs[frame] !== undefined) ? meshs[frame] : -1;

        const buffer = g_models[id][frame];
        if (null == buffer) debugger

        let subMesh: RenderingSubMesh = null!;
        if (mesh >= 0) {
            subMesh = EFKModel.m_meshs[mesh];
        }
        else {
            const attrs: gfx.Attribute[] = [];
            material.passes[0].shaderInfo.attributes.forEach(attr => {
                attrs.push(new gfx.Attribute(attr.name, attr.format, attr.isNormalized, attr.stream, attr.isInstanced, attr.location));
            })
            subMesh = new RenderingSubMesh([buffer.vertex], ModelVertexEx, gfx.PrimitiveMode.TRIANGLE_LIST, buffer.indice);

            mesh = EFKModel.m_meshs.length;
            EFKModel.m_meshs.push(subMesh);
            EFKModel.m_model_meshs[id].push(mesh);
        }

        return EFKModel.checkModel(mesh, subMesh, material, buffer.ia.drawInfos[0]);
    }

    private static checkModel(mesh_id: number, mesh_obj: RenderingSubMesh, material: Material, info: gfx.DrawInfo): renderer.scene.Model {
        let model = null as renderer.scene.Model;
        const cache = EFKModel.m_model_cache[mesh_id] = (EFKModel.m_model_cache[mesh_id] || []);
        for (let ii = 0; ii < cache.length; ++ii) {
            let mid = cache[ii];
            if (EFKModel.m_model_used[mid]) {
                continue;
            }
            model = EFKModel.m_models[mid];
            EFKModel.m_model_used[mid] = true;
            break;
        }

        if (null == model) {
            model = director.root.createModel(renderer.scene.Model);
            model.initSubModel(0, mesh_obj, material);

            const mid = EFKModel.m_models.length;
            EFKModel.m_models.push(model);

            EFKModel.m_model_cache[mesh_id] = EFKModel.m_model_cache[mesh_id] || [];
            EFKModel.m_model_cache[mesh_id].push(mid);

            EFKModel.m_model_used[mid] = true;
        }
        else {
            model.setSubModelMaterial(0, material);
        }

        model.subModels[0].inputAssembler.drawInfo = info;
        model.priority = EFKModel.m_priority++;

        return model;
    }

    public static reset() {
        EFKModel.m_typeid_used = [];
        EFKModel.m_model_used = [];
        EFKModel.m_priority = 0;
    }
}

File: extensions/effekseer/engine/native/effekseer/EffekseerCocosCreator/framework/efk_model.cpp

#include "./efk_model.h"

using namespace cc;

const ccstd::vector<gfx::Attribute> ModelVertexEx{
    gfx::Attribute{ "Input_Pos", gfx::Format::RGB32F },       // Pos      -> vec3 + 12
    gfx::Attribute{ "Input_Normal", gfx::Format::RGB32F },    // Normal   -> vec3 + 12
    gfx::Attribute{ "Input_Binormal", gfx::Format::RGB32F },  // Binormal -> vec3 + 12
    gfx::Attribute{ "Input_Tangent", gfx::Format::RGB32F },   // Tangent  -> vec3 + 12
    gfx::Attribute{ "Input_UV", gfx::Format::RG32F },         // UV1      -> vec2 + 8
    gfx::Attribute{ "Input_Color", gfx::Format::RGBA32F },    // Color    -> vec4 + 16
                                                              //                  = 72
};

const int ModelStrip = 72;

struct ModelInfo {
    gfx::Buffer*  vertex;
    gfx::Buffer*  indice;
    gfx::DrawInfo ia;
};

static std::unordered_map<int32_t, std::vector<ModelInfo>> g_models;

struct SpriteInfo {
    gfx::Buffer* vertex;
    gfx::Buffer* indice;
};

static std::unordered_map<int32_t, SpriteInfo> m_buffers;

static std::vector<RenderingSubMesh*>                    m_meshs;
static std::unordered_map<int32_t, std::vector<int32_t>> m_typeid_meshs;
static std::unordered_map<int32_t, bool>                 m_typeid_used;
static std::unordered_map<int32_t, std::vector<int32_t>> m_model_meshs;
static std::unordered_map<int32_t, std::vector<int32_t>> m_model_cache;

static std::vector<scene::Model*>        m_models;
static std::unordered_map<int32_t, bool> m_model_used;

static int32_t m_priority = 0;

void EFKModelN::reset() {
    m_typeid_used.clear();
    m_model_used.clear();
    m_priority = 0;
}

void EFKModelN::addSprite(uint32_t stride, std::vector<uint8_t> vertecies) {
    gfx::Device* device = gfx::Device::getInstance();
    gfx::Buffer* vertex = nullptr;

    auto itr = m_buffers.find(stride);
    if (itr != m_buffers.end()) {
        auto sprite = itr->second;
        vertex      = sprite.vertex;
        if (sprite.vertex->getSize() < vertecies.size()) {
            sprite.vertex->resize(vertecies.size());
        }
    }
    else {
        vertex = device->createBuffer({ gfx::BufferUsageBit::VERTEX | gfx::BufferUsageBit::TRANSFER_DST, gfx::MemoryUsageBit::HOST | gfx::MemoryUsageBit::DEVICE, (uint32_t)vertecies.size(), stride });

        uint32_t     count  = 3;  // (vertexBuffer.size / stride) / 4 * 2 * 3;
        gfx::Buffer* indice = device->createBuffer({ gfx::BufferUsageBit::INDEX | gfx::BufferUsageBit::TRANSFER_DST, gfx::MemoryUsageBit::DEVICE, (uint32_t)(count * sizeof(uint16_t)), sizeof(uint16_t) });

        m_buffers.insert(std::make_pair(stride, SpriteInfo{
                                                    vertex,
                                                    indice }));
    }

    vertex->update(vertecies.data(), vertecies.size());
}

void EFKModelN::resizeIndice(gfx::Buffer* indice, uint32_t count) {
    if (indice->getSize() >= count * sizeof(uint16_t)) {
        return;
    }

    indice->resize(count * sizeof(uint16_t));
    auto data = new uint16_t[count];

    auto element = 0;
    for (auto i = 0; i < count;) {
        auto baseIdx = 4 * element;
        data[i++]  = baseIdx;
        data[i++]  = baseIdx + 1;
        data[i++]  = baseIdx + 2;
        data[i++]  = baseIdx + 3;
        data[i++]  = baseIdx + 2;
        data[i++]  = baseIdx + 1;
        element++;
    }
    indice->update(data, count * sizeof(uint16_t));
}

scene::Model* EFKModelN::drawSprite(uint32_t stride, uint32_t offset, uint32_t spriteCount, cc::Material* material, uint32_t type) {
    auto fisrtVertex = offset / stride;
    auto fisrtIndex  = fisrtVertex / 4 * 6;

    auto buffer = m_buffers[stride];
    EFKModelN::resizeIndice(buffer.indice, fisrtIndex + spriteCount * 6);

    auto iaInfo = gfx::DrawInfo{
        spriteCount * 4,
        fisrtVertex,
        spriteCount * 6,
        fisrtIndex
    };

    auto mesh = -1;
    auto itr  = m_typeid_meshs.find(type);
    if (itr != m_typeid_meshs.end()) {
        m_typeid_meshs.insert(std::make_pair(type, std::vector<int32_t>{}));
    }

    auto& meshs = m_typeid_meshs[type];
    for (auto i = 0; i < meshs.size(); ++i) {
        auto idx = meshs[i];
        if (!m_typeid_used[idx]) {
            m_typeid_used.insert(std::make_pair(idx, true));
            mesh = idx;
            break;
        }
    }

    RenderingSubMesh* subMesh = NULL;
    if (mesh >= 0) {
        subMesh = m_meshs[mesh];
        subMesh->setDrawInfo(iaInfo);
    }
    else {
        std::vector<gfx::Attribute> attrs;

        scene::Pass* pass      = material->getPasses().get()->at(0).get();
        auto         attrs_src = pass->getShaderInfo()->attributes;
        for (auto attr : attrs_src) {
            attrs.push_back(gfx::Attribute{ attr.name, attr.format, attr.isNormalized, attr.stream, attr.isInstanced, attr.location });
        }

        subMesh = new RenderingSubMesh(std::vector<gfx::Buffer*>{ buffer.vertex }, attrs, gfx::PrimitiveMode::TRIANGLE_LIST, buffer.indice, nullptr, false);
        subMesh->addRef();

        mesh = m_meshs.size();
        m_meshs.push_back(subMesh);

        m_typeid_meshs[type].push_back(mesh);
        m_typeid_used[mesh] = true;
    }

    return EFKModelN::checkModel(mesh, subMesh, material, iaInfo);
}

void EFKModelN::addModel(int32_t key, std::vector<uint8_t> vertecies, std::vector<uint8_t> indices, uint32_t frame) {
    gfx::Device* device = gfx::Device::getInstance();

    auto itr = g_models.find(key);
    if (itr != g_models.end()) {
        g_models.insert(std::make_pair(key, std::vector<ModelInfo>{}));
    }

    if (g_models[key].size() > frame) {
        auto& buffer = g_models[key][frame];
        if (buffer.vertex->getSize() < vertecies.size()) {
            buffer.vertex->resize(vertecies.size());
            buffer.indice->resize(indices.size());
        }
    }
    else {
        auto vertex = device->createBuffer({ gfx::BufferUsageBit::VERTEX | gfx::BufferUsageBit::TRANSFER_DST, gfx::MemoryUsageBit::HOST | gfx::MemoryUsageBit::DEVICE, (uint32_t)vertecies.size(), ModelStrip });
        auto indice = device->createBuffer({ gfx::BufferUsageBit::INDEX | gfx::BufferUsageBit::TRANSFER_DST, gfx::MemoryUsageBit::DEVICE, (uint32_t)indices.size(), sizeof(uint32_t) });

        vertex->addRef();
        indice->addRef();

        g_models[key].resize(frame + 1);
        g_models[key][frame] = {
            vertex,
            indice,
            { vertex->getCount(),
              0,
              indice->getCount(),
              0 }
        };
    }

    g_models[key][frame].vertex->update(vertecies.data(), vertecies.size());
    g_models[key][frame].indice->update(indices.data(), indices.size());
}

void EFKModelN::removeModel(int32_t key) {
    if (g_models.find(key) == g_models.end()) {
        return;
    }

    auto& frames = g_models[key];
    for (auto frame : frames) {
        frame.vertex->release();
        frame.indice->release();
    }
    g_models.erase(key);

    auto& meshs = m_model_meshs[key];
    for (auto mesh : meshs) {
        m_meshs[mesh]->release();

        auto& models = m_model_cache[mesh];
        for (auto model : models) {
            m_models[model]->release();
        }
        m_model_cache.erase(mesh);
    }
    m_model_meshs.erase(key);
}

scene::Model* EFKModelN::drawModel(int32_t id, uint32_t frame, Material* material, uint32_t type) {
    auto itr = m_model_meshs.find(id);
    if (itr == m_model_meshs.end()) {
        m_model_meshs.insert(std::make_pair(id, std::vector<int32_t>{}));
    }

    auto& meshs = m_model_meshs[id];
    auto  mesh  = (meshs.size() > frame && meshs[frame] != 0) ? meshs[frame] : -1;

    auto              buffer  = g_models[id][frame];
    RenderingSubMesh* subMesh = nullptr;

    if (mesh >= 0) {
        subMesh = m_meshs[mesh];
    }
    else {
        subMesh = new RenderingSubMesh({ buffer.vertex }, ModelVertexEx, gfx::PrimitiveMode::TRIANGLE_LIST, buffer.indice);
        subMesh->addRef();

        mesh = m_meshs.size();
        m_meshs.push_back(subMesh);
        m_model_meshs[id].push_back(mesh);
    }

    return EFKModelN::checkModel(mesh, subMesh, material, buffer.ia);
}

scene::Model* EFKModelN::checkModel(int32_t mesh_id, RenderingSubMesh* mesh_obj, Material* material, gfx::DrawInfo& info) {
    scene::Model* model = nullptr;

    auto itr = m_model_cache.find(mesh_id);
    if (itr == m_model_cache.end()) {
        m_model_cache.insert(std::make_pair(mesh_id, std::vector<int32_t>{}));
    }

    auto& cache = m_model_cache[mesh_id];
    for (auto ii = 0; ii < cache.size(); ++ii) {
        auto mid = cache[ii];
        if (m_model_used[mid]) {
            continue;
        }
        model             = m_models[mid];
        m_model_used[mid] = true;
        break;
    }

    if (nullptr == model) {
        model = Root::getInstance()->createModel<scene::Model>();
        model->addRef();
        model->initSubModel(0, mesh_obj, material);

        auto mid = m_models.size();
        m_models.push_back(model);
        m_model_cache[mesh_id].push_back(mid);

        m_model_used[mid] = true;
    }
    else {
        model->setSubModelMaterial(0, material);
    }

    model->getSubModels()[0]->getInputAssembler()->setDrawInfo(info);
    model->setPriority(m_priority++);

    return model;
}
